### ISSUE TITLE
Fix S3168 FN: Async implementation of non-async interfaces

### ISSUE THREAD

**Issue by qhris** [SonarSource/sonar-dotnet#9696]
### Description

Async implementations of non-async interface methods are not properly reported as an issue.

### Repro steps

Minimal reproduction code:

```cs
public interface ITestReproduction
{
    void TestMethod();
}

public class TestReproduction : ITestReproduction
{
    async void TestMethod() // FN: Not reported as an issue.
    {
        await Task.Delay(1000);
    }
}
```

### Expected behavior

The above code should report an issue with the async void implementation of the interface.

### Actual behavior

Nothing is reported.

### Known workarounds

Nothing.

### Related information

* C# 12
* .NET 8.0
* Rider 2024.3 with SonarLint 10.12.0.79769
* Operating System Linux

**Comment by Tim-Pohlmann** [comment]
Hi @qhris!
This case is excluded because it cannot be fixed without changing the interface or making the method non-async. I will update [rspec](https://sonarsource.github.io/rspec/#/rspec/S3168) to clarify this.
Let me know if you need more clarification.

**Comment by Tim-Pohlmann** [comment]
Here is the rspec PR:
https://github.com/SonarSource/rspec/pull/4547

**Comment by qhris** [comment]
Seems a bit strange to me, the fix is pretty clear to me: make the method synchronous. @Tim-Pohlmann 

The entire `async void` really needs to be reported regardless in my opinion.
It is a pretty nasty footgun which can bring down the entire application (and did in our case, by a segfault).

Quoting from Microsoft [here](<https://learn.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming#avoid-async-void>):

> Async void methods have different error-handling semantics. When an exception is thrown out of an async Task or async Task<T> method, that exception is captured and placed on the Task object. With async void methods, there is no Task object, so any exceptions thrown out of an async void method will be raised directly on the SynchronizationContext that was active when the async void method started.

Should this be a separate rule then? Catching stuff like this is what we want to use the sonar scanner for in the first place.

**Comment by Tim-Pohlmann** [comment]
@qhris That is a Very valid point; thanks for your input!
I created an issue for a [new rule idea](https://github.com/SonarSource/sonar-dotnet/issues/9711). Please review it and let me know if this is what you are looking for.
Thanks for challenging our decisions; this is how we grow from user feedback and improve our products to better suit your needs!
